### ISSUE TITLE
Add option to arrow-body-style to disallow braces for single-statement functions

### ISSUE THREAD

**Issue by Kerrick** [eslint/eslint#10836]
**What rule do you want to change?**

[`arrow-body-style`](https://eslint.org/docs/rules/arrow-body-style)

**Does this change cause the rule to produce more or fewer warnings?**

More

**How will the change be implemented? (New option, new default behavior, etc.)?**

I'd be happy if it was a new option _or_ if `as-needed` changed its behavior.

**Please provide some example code that this change will affect:**

```js
let foo = () => { bar(); };
```

I'd like to propose 
**What does the rule currently do for this code?**

Currently, that code is considered **correct** for `as-needed`.

**What will the rule do after it's changed?**

If the proposed enhancement is a change to `as-needed`, that code would become incorrect.

If the proposed enhancement is a new option, the new option would function exactly like `as-needed` _except_ that code would become incorrect.

---

While it is technically different from `let foo = () => bar();` because the return value is `undefined` for the with-braces version but the brace-less version returns the return value of `bar()`, in day-to-day code it often doesn't make a difference. It is often stylistically cleaner _not_ to have the braces, especially when using an auto-formatter like Prettier.js that multi-lines every brace-including arrow function.

Examples of code that don't care whether the braces are there (`return undefined` implicit) or not (return the statement's value):

```js
imgTag.onload = () => myApp.showSuccess();
```

```js
Ember.run(() => this.set('foo', 'bar'));
```

```js
QUnit.assert.throws(() => myApp.shouldThrow());
```

**Comment by ilyavolodin** [comment]
How would we identify the cases where `return undefined` matters and when it doesn't?

**Comment by Kerrick** [comment]
ESLint doesn’t, the programmer does. They tell the linter (and other developers) that the return value matters by returning it. They already are using braces anyways, and it’s much more explicit.

**Comment by not-an-aardvark** [comment]
Thanks for the proposal. If I'm understanding correctly:

* In cases where the return value of a single-statement arrow function doesn't matter, you consider it stylistically cleaner to use an implicit-return arrow function.
* In cases where the return value of a single-statement arrow function does matter, the suggestion is to use an extra `return;` statement in the block to indicate that the return value doesn't matter. This is a matter of opinion, but my perspective is that a redundant return statement is stylistically much less clean.

On balance, it's not obvious to me that there's a clear value-add here. The reasoning for adding the redundant return statement is that it's "much more explicit that way", but I don't see how that's true; ending a function block without returning clearly communicates that `undefined` is returned to any reader familiar with the language. On the other hand, returning a value by omitting a block for the sake of stylistic cleanliness seems like it could cause confusion due to a lack of explicitness; when reading and refactoring, it might be difficult to tell which return values are actually used, and which return values are only present to make the code look nice.

**Comment by Kerrick** [comment]
> On the other hand, returning a value by omitting a block for the sake of stylistic cleanliness seems like it could cause confusion due to a lack of explicitness; when reading and refactoring, it might be difficult to tell which return values are actually used, and which return values are only present to make the code look nice.

In the same vein, returning `undefined` by including a block because braces were typed because "other functions have braces so why not this one" seems like it could cause confusion due to a lack of explicitness; when reading and refactoring, it might be difficult to tell which `undefined` return values are actually used, and which `undefined` return values are only present because the developer "accidentally" used braces.

**Comment by not-an-aardvark** [comment]
> Nope, the suggestion to use an extra return is only in cases where the implicit undefined return value of a single-statement arrow function with a block matters.

Yes, that's what I meant -- sorry, my wording was a bit unclear. Thanks for clarifying.

> In the same vein, returning `undefined` by including a block because braces were typed ("because other functions have braces so why not this one") seems like it could cause confusion due to a lack of explicitness; when reading and refactoring, it might be difficult to tell which `undefined` return values are actually used, and which `undefined` return values are only present because the developer "accidentally" used braces.

I'm not sure I follow. Could you elaborate on what you mean by "accidentally" using braces?

---

Here's an example to demonstrate what I mean about readability:

```js
// Sample A
doSomethingWithACallback(() => foo());
```

```js
// Sample B
doSomethingWithACallback(() => {
    foo();
});
```

```js
// Sample C
doSomethingWithACallback(() => {
    foo();
    bar();
});
```

```js
// Sample D
doSomethingWithACallback(() => {
    bar();
    return foo();
});
```

Suppose my code looks like Sample A, and I want to update the callback so that `bar()` gets called in addition to `foo()`. From just looking at the code, it's not clear whether the return value of `foo()` is used in `doSomethingWithACallback`. If the return value is unused, then the correct update would look like Sample C. If the return value is used, then the correct update would look like Sample D (to ensure that the same value is returned as before). In order to figure out the correct refactoring, I might have to look at the implementation of `doSomethingWithACallback` and/or `foo`.

On the other hand, if my code initially looks like Sample B, then it's clear that the return value of `foo` is not being used, because it never leaves the callback. So I can immediately tell that Sample C would be the correct refactoring, without needing to look at the implementation of `doSomethingWithACallback` or `foo`.

In other words, in the case where a return value is unused, Sample B is more "explicit" than Sample A in that it provides more clarity about how the system behaves. The intent is clear based on the presence of the block, despite the fact that the `return` keyword doesn't appear anywhere in the text.

**Comment by ljharb** [comment]
One common pattern in React is:
```jsx
<div onClick=(() => { this.something(); }} />
```

This is *critically necessary*, because if `this.something()` returns false, it will stop event propagation. In other words, the braces are absolutely necessary, and converting it to not use the braces is **incorrect** and can and will break sites.

I agree with the above; explicit use of braces is the signal that the return value must be `undefined`, not the presence or absence of the `return` keyword.

**Comment by ilyavolodin** [comment]
I'm with @not-an-aardvark here. I think what you are proposing actually lowers readability, instead of improving it. It's always better to be explicit.

**Comment by Kerrick** [comment]
Don’t you consider using the return keyword to signify that you care about the return value of block arrow functions more explicit?

**Comment by not-an-aardvark** [comment]
I don't think explicitness on its own should be a first-class concern when deciding how to write code. Explicitness is useful when:

* it disambiguates two plausible meanings for a piece of code, making it clear to the reader which one was intended
* it adds intentional overhead to a discouraged pattern to prevent users from accidentally using that pattern unless they're sure they know what they're doing

Both of these are possible benefits of being explicit, but I don't think it's useful to use a code pattern just for the sake of being explicit unless it helps to accomplish some other readability goal.

**Comment by platinumazure** [comment]
If the ESLint team does not believe your proposal is worth including in core, remember that you are welcome to create a custom rule that does what you want. We have to take on maintenance burden with every change request we implement, so we want to only implement changes that benefit a wide cross section of users. But, we designed support for custom rules and plugins specifically so that our design decisions for core rules don't get in your way: you can always create a custom rule for your use case.

Also, we are happy to help you in the creation of said custom rule, if you need help.

**Comment by zakkudo** [comment]
I'm going to do a follow up on this as a I believe it glosses over a lot of context for this.  The current rule encourages arbitrary refactoring to play with a new toy and in no way encourages cleaner or clearer coding.

In python a rule like this for fat arrows (lambda functions) would only be allowed for a single line to add clarity that it is that kind of statement.  `x => x.id` is a good use of no parenthesis. 
```
x => (
    something;
    something;
    something;
)
```
in my opinion is not and we already have previous examples in other languages.  We don't need it as a default, but I do recommend it as a configurable.

https://dbader.org/blog/python-lambda-functions

**Comment by not-an-aardvark** [comment]
@zakkudo I'm not sure I follow; are you arguing that people shouldn't use multi-line arrow functions in JavaScript because complex lambdas are discouraged in Python? It's not clear to me that this effect carries over. I think complex lambdas in Python are discouraged because of two aspects which are not present in JavaScript:

* Python doesn't support having multiple statements in lambda expressions
* In Python, functions like `map` and `filter` are globally available and accept two arguments (the iterable and a function), rather than being defined on iterables and only accepting the function as an argument. As a result, `map` and `filter` resemble prefix operators rather than infix operators. This makes it significantly more difficult for a human to parse code that uses `map`/`filter` in Python, because a lambda is located far away from the name of the callee, making it hard to determine what the lambda means in context.

**Comment by zakkudo** [comment]
The larger context is, instead of adding options to `@private`, we should probably consider add options to the other rules to accommodate private since it is an edge case.

**Comment by zakkudo** [comment]
Bleh.  I think I may be commenting on the wrong github issue...

**Comment by zakkudo** [comment]
The point is that this use case sounds like the python use case for readable code.  I think an option might be appropriate.  The current option can stay is it is, but as it is it falls more on the forcing people to play with a new toy side rather than being a useful tool for normalizing coding styles.

**Comment by ljharb** [comment]
This issue isn't about readability - `x => { y(); }` and `x => y()` have very different semantics and are very different things. Both are equally readable in that they precisely convey the desired semantics - in the former, a return value of `undefined`, and in the latter, a return value of "the return value of `y`".

**Comment by zakkudo** [comment]
Yep get the wrong ticket!  sorry.  I'll make a different issue.  Sorry guys...

**Comment by Kerrick** [comment]
Sure, it’s not about readability. It’s about semantics and explicitness. In the same way I want eslint to be able to enforce == vs ===, which is about semantics and explicitness, I would love an option to see this enforced.

**Comment by ljharb** [comment]
@platinumazure my comment is not off-topic; it's quite related to this issue - please unhide it.

@Kerrick the different is that `==` vs `===` is likely a bug; the likely bug is *omitting* the braces, unintentionally exposing a return value, which unintentionally becomes part of the contract of your function.

**Comment by not-an-aardvark** [comment]
@ljharb I had marked the conversation as off-topic because it appears that it the argument in it was intended for a different issue. I've unhidden your comment.

**Comment by Kerrick** [comment]
> @Kerrick the different is that == vs === is likely a bug; the likely bug is omitting the braces, unintentionally exposing a return value, which unintentionally becomes part of the contract of your function.

You raise a really good point here. I had considered it to be more about accidentally _not_ returning a value, but you're probably right that the more likely accident is to return a value you didn't intend.
